RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletGeneratorKernelsImpl.h

```diff
@@ -499,43 +694,139 @@ __global__ void kernel_sharedHitCleaner(TrackingRecHit2DSOAView const *__restric
+                                      HitContainer const *__restrict__ ptuples,
+                                      TkSoA const *__restrict__ ptracks,
+                                      Quality *__restrict__ quality,
+                                      uint16_t nmin,
+                                      bool dupPassThrough,
+                                      CAHitNtupletGeneratorKernelsGPU::HitToTuple const *__restrict__ phitToTuple) {
+  // quality to mark rejected
+  auto const reject = pixelTrack::Quality::loose;
+  /// min quality of good
+  auto const good = pixelTrack::Quality::strict;
+
+  auto &hitToTuple = *phitToTuple;
+  auto const &foundNtuplets = *ptuples;
+  auto const &tracks = *ptracks;
+
+  int first = blockDim.x * blockIdx.x + threadIdx.x;
+  for (int idx = first, ntot = hitToTuple.nOnes(); idx < ntot; idx += gridDim.x * blockDim.x) {
+    if (hitToTuple.size(idx) < 2)
+      continue;
+
+    float mc = 10000.f;
+    uint16_t im = 60000;
+    uint32_t maxNh = 0;
+
+    // find maxNh
+    for (auto it = hitToTuple.begin(idx); it != hitToTuple.end(idx); ++it) {
+      if (quality[*it] <= good)
+        continue;
+      uint32_t nh = foundNtuplets.size(*it);
+      maxNh = std::max(nh, maxNh);
+    }
+
+    // only triplets
+    if (maxNh != 3)
+      continue;
+
+    // for triplets choose best tip!  (should we first find best quality???)
+    for (auto ip = hitToTuple.begin(idx); ip != hitToTuple.end(idx); ++ip) {
+      auto const it = *ip;
+      if (quality[it] >= good && std::abs(tracks.tip(it)) < mc) {
+        mc = std::abs(tracks.tip(it));
+        im = it;
+      }
+    }
+
+    if (60000 == im)
+      continue;
+
+    // mark worse ambiguities
+    for (auto ip = hitToTuple.begin(idx); ip != hitToTuple.end(idx); ++ip) {
+      auto const it = *ip;
+      if (quality[it] > reject && it != im)
+        quality[it] = reject;  //no race:  simple assignment of the same constant
     }
 
-    if (maxNh > 3)
+  }  // loop over hits
+}
+
+__global__ void kernel_simpleTripletCleaner(
+    TrackingRecHit2DSOAView const *__restrict__ hhp,
+    HitContainer const *__restrict__ ptuples,
+    TkSoA const *__restrict__ ptracks,
+    Quality *__restrict__ quality,
+    uint16_t nmin,
+    bool dupPassThrough,
+    CAHitNtupletGeneratorKernelsGPU::HitToTuple const *__restrict__ phitToTuple) {
+  // quality to mark rejected
+  auto const reject = pixelTrack::Quality::loose;
+  /// min quality of good
+  auto const good = pixelTrack::Quality::loose;
+
+  auto &hitToTuple = *phitToTuple;
+  auto const &foundNtuplets = *ptuples;
+  auto const &tracks = *ptracks;
+
+  int first = blockDim.x * blockIdx.x + threadIdx.x;
+  for (int idx = first, ntot = hitToTuple.nOnes(); idx < ntot; idx += gridDim.x * blockDim.x) {
+    if (hitToTuple.size(idx) < 2)
       continue;
-    // for triplets choose best tip!
+
+    float mc = 10000.f;
+    uint16_t im = 60000;
```

Comment:
here and elsewhere, could you use something like `constexpr uint16_t notFound = 60000;` ?